RSDK-5830 Add static FFmpeg libraries for Linux x86_64

bazile-clyde opened the pull request:
# Description 
This PR adds static FFmpeg libraries for linux/amd64. This does _not_ include support for 32-bit OSes. That and darwin support will be done in later tickets.

# Testing
I tested this on my laptop (Linux x86_64) for the new functionality and a Raspberry Pi (Linux aarch64) for regression testing.

bazile-clyde commented on Makefile:
```diff
@@ -136,20 +136,20 @@ $(FFMPEG_ROOT):
+	FFMPEG_OPTS += --enable-encoder=h264_v4l2m2m
+	FFMPEG_OPTS += --enable-v4l2-m2m
+endif
+ffmpeg: $(FFMPEG_ROOT)
```
@seanavery running `make ffmpeg` for darwin builds may just work. This command will only keep the archive files and leave everything else alone. You'll then have to update the `LDFLAGS` in `avcodec.go`, `avutils.go`, and `avlog.go`.

bazile-clyde commented on web/server/entrypoint_linux.go:
```diff
@@ -1,17 +1,23 @@
-//go:build cgo && linux && !android
+//go:build cgo && linux && !arm
```
@abe-winter I'm leaving the 32-bit alone for now so I'm not exactly sure why the 32-bit tests are failing but it doesn't look like it has anything to do with these changes, especially considering main has been acting janky. I'll keep looking into though until I get a definitive answer.

abe-winter replied:
(32 bit passing now, resolving this)

bazile-clyde commented on web/server/entrypoint_linux.go:
```diff
@@ -1,17 +1,23 @@
 	var streamConfig gostream.StreamConfig
-	// TODO(RSDK-5916): support v4l2m2m codec on arm64/rpi
-	streamConfig.VideoEncoderFactory = x264.NewEncoderFactory()
+	if avcodec.EncoderIsAvailable(h264.V4l2m2m) {
```
This adds back the RPi HW encoder, but I left that ticket open and added it to the epic.

seanavery replied:
`EncoderIsAvailable` check is nice and simple! 

[q] Are we sure we want to add back in without addressing edge case issues first?

abe-winter commented on Makefile:
```diff
@@ -138,20 +138,20 @@ $(FFMPEG_ROOT):
+	FFMPEG_OPTS += --enable-encoder=h264_v4l2m2m
+	FFMPEG_OPTS += --enable-v4l2-m2m
+endif
+ffmpeg: $(FFMPEG_ROOT)
```
unrelated -- if FFMPEG_ROOT step takes a long time, you can change it to
```
git clone https://github.com/FFmpeg/FFmpeg.git --depth 1 --branch release/6.1
```

this will fetch 20mb instead of hundreds

bazile-clyde replied:
Nice. Done!

abe-winter commented on Makefile:
```diff
@@ -138,20 +138,20 @@ $(FFMPEG_ROOT):
-	rm -rf $(FFMPEG_H264_PREFIX)/lib/pkgconfig
-	rm -rf $(FFMPEG_H264_PREFIX)/share
+	# only keep archive files
+	find $(FFMPEG_PREFIX)/* -type d ! -wholename $(FFMPEG_PREFIX)/lib | xargs rm -rf
```
why is this cleanup step necessary?

bazile-clyde replied:
`make install` will add source and pkg-config files to `FFMPEG_PREFIX`, and as far as I can tell, there's no way to prevent that. We don't need pkg-config files, and the source files are shared among the architectures and are in the directory above. I just added a comment too, to clarify.

abe-winter reviewed (commented):
looks good -- pushing these comments, doing one last pass to look at entrypoint_*.go tagging, then :heavy_check_mark: 

very happy rimage is no longer `no_cgo`

abe-winter commented on gostream/ffmpeg/include/libavcodec/ac3_parser.h:
why is ffmpeg cloned from git, but libavcodec is checked in?

bazile-clyde replied:
This rule is to make it easier to build ffmpeg's static library. We will rarely use it, but it's helpful to keep upgrading and adding support for different OSes consistently. We can just run `make ffmpeg` on our chosen arch, and it'll create and copy the archive files for that arch to where they need to be. 

We need to link against the archive files themselves, so I checked those in. I didn't want to force developers to have to build FFmpeg every time they clone the repo or risk rebuilding it wrongly anytime we want to upgrade or add support.

Does that answer the question?

abe-winter reviewed (approved):
looks good

entrypoint platform selection logic here if helpful:
https://docs.google.com/spreadsheets/d/1C3lwZ2Aa4HmwuCCD12YNKfvw_X9wH4F3_-wVKU12nzs/

seanavery commented on Makefile:
```diff
@@ -134,24 +134,23 @@ license-check:
-		--prefix=$(FFMPEG_H264_PREFIX) \
-		--enable-pic
+FFMPEG_PREFIX ?= $(shell realpath .)/gostream/ffmpeg/$(shell uname -s)-$(shell uname -m)
+FFMPEG_OPTS = --disable-programs --disable-doc --disable-everything --prefix=$(FFMPEG_PREFIX) --enable-pic --disable-autodetect
```
[info] I had not looked into the `PIC` linking flag before. Found GNU reference [here](https://gcc.gnu.org/onlinedocs/gcc/Code-Gen-Options.html#index-fpic).

bazile-clyde replied:
Great catch! I removed this since I don't think we need it. Everything works fine without it.

seanavery commented on gostream/.gitignore:
```diff
@@ -1,56 +1,2 @@
-
-etc/test_keys
+# Keep FFmpeg static libraries 
+!**/ffmpeg/**/*.a
```
nice cleanup here!

seanavery commented on gostream/codec/h264/ffmpeg/Linux-aarch64/include/libavcodec/vaapi.h:
[q] Are we removing this because we can safely assume no display integrations?

bazile-clyde replied:
Kind of. I think this was automatically removed when I added `--disable-autodetect` to FFmpeg's build options. That stopped including this lib and a couple others automatically if your system had it. The issue that solved was whenever I would run `make ffmpeg` on my `x86` desktop it would include libs like this in the build, but whenever I'd run that same static build in docker it'd fail because those libs don't exist. We don't need them yet, so no need to include them.

seanavery commented on gostream/ffmpeg/avlog/avlog.go:
```diff
@@ -0,0 +1,53 @@
+//#include <libavutil/log.h>
+import "C"
+
+const (
```
[q] It looks like we are missing a couple of macros here. Do you think we should add these?
- [AV_LOG_PRINT_LEVEL](https://ffmpeg.org/doxygen/3.0/group__lavu__log.html#ga8b447fa5f9d7e243be4634878ed1a0ca)   2
- [AV_LOG_SKIP_REPEATED](https://ffmpeg.org/doxygen/3.0/group__lavu__log.html#ga6cdf5cd331b17e80e8308124f05a6db8)   1

seanavery replied:
[q] Is it possible and/or good practice to use macros directly from the the headerfile (cgo import), instead of defining our own data structures for them?

bazile-clyde replied:
I'm not using either of those, so I didn't include them. I only added the rest of the log levels so we wouldn't have to look them up later. I was also playing around with the different levels and decided to leave them in. I generally tried to avoid importing things that aren't tested or used in prod.

From what I've seen, importing macros from C isn't really a thing. People have been getting around it by defining a new C function that wraps the macro. That's how it's done in avutil.go, where `av_err2str` is a macro.
```
//#include <libavutil/error.h>
//#include <stdlib.h>
// static const char *error2string(int code) { return av_err2str(code); }
import "C"
...
// ErrorFromCode returns an error from the given code
func ErrorFromCode(code int) error {
	if code >= 0 {
		return nil
	}

	return errors.New(C.GoString(C.error2string(C.int(code))))
}
```
Because you can't call a function and use its value to instantiate a constant, I just added the values myself. But to answer your question, yes, I think it'd be better practice to use the macros directly somehow. I'm just not sure how at the moment. If I find out, or if you know of a way, I'd be happy to update this!

seanavery commented on gostream/ffmpeg/avlog/avlog.go:
```diff
@@ -0,0 +1,53 @@
+
+const (
+	// LogQuiet Print no output.
+	LogQuiet = (iota * 8) - 8
```
Nice!

seanavery reviewed (approved):
Done with first pass, and LGTM! 
Only concern is adding back in the h264 hardware encoder without further testing.

bazile-clyde commented on Makefile:
```diff
@@ -138,20 +138,20 @@ $(FFMPEG_ROOT):
+FFMPEG_PREFIX ?= $(shell realpath .)/gostream/ffmpeg/$(shell uname -s)-$(shell uname -m)
+FFMPEG_OPTS = --disable-programs --disable-doc --disable-everything --prefix=$(FFMPEG_PREFIX) --enable-pic --disable-autodetect
+ifeq ($(shell uname -m),aarch64)
+	FFMPEG_OPTS += --enable-encoder=h264_v4l2m2m
```
Because we only support hardware encoding on RPis, there's no need to add this econder to non-aarch64 archs. I added a comment to clarify.

bazile-clyde commented:
> [q] Are we sure we want to add back in without addressing edge case issues first?

@seanavery great point. I added the TODO back in and will look at the flakiness and the referenced ticket. I also added a simple test for `EncoderIsAvailable` in `avcodec_test.go` to ensure we don't lose that functionality.
